pkg/logging/logging.go

```diff
@@ -0,0 +1,189 @@
+		defaultVerbosity, _ = strconv.Atoi(verbosityFlag.Value.String())
+	} else {
+		defaultVerbosity = 0
+	}
+	return &FilteredLogger {
+		logContext:            log.NewContext(logger).With("component", defaultComponent),
+		filterLevel:           defaultLogLevel,
+		currentLogLevel:       defaultLogLevel,
+		verbosityLevel:        defaultVerbosity,
+		currentVerbosityLevel: defaultVerbosity,
+	}
+}
+
+type NullLogger struct{}
+
+func (n NullLogger) Log(params ...interface{}) error { return nil }
+
+var loggers = make(map[string]*FilteredLogger)
+var defaultComponent = ""
+var defaultVerbosity = 0
+
+func Logger(component string) *FilteredLogger {
+	if _, ok := loggers[component]; !ok {
+		logger := log.NewLogfmtLogger(os.Stderr)
+		log := MakeLogger(logger).With("component", component)
+		loggers[component] = log
+	}
+	return loggers[component]
+}
+
+func (l *FilteredLogger) SetLogger(logger log.Logger) *FilteredLogger {
+	l.logContext = log.NewContext(logger)
+	return l
+}
+
+type LogError struct {
+	message string
+}
+
+func (e LogError) Error() string {
+	return e.message
+}
+
+func (l FilteredLogger) Log(params ...interface{}) error {
+	// messages should be logged if any of these conditions are met:
+	// The log filtering level is debug
+	// The log filtering level is info and verbosity checks match
+	// The log message priority is warning or higher
+	force := (l.filterLevel == DEBUG) || (l.currentLogLevel >= WARNING)
+
+	if force || (l.filterLevel == INFO &&
+		(l.currentLogLevel == l.filterLevel) &&
+		(l.currentVerbosityLevel <= l.verbosityLevel)) {
+			logParams := make([]interface{}, 0)
+
+			if l.currentVerbosityLevel > 1 {
+				_, fileName, lineNumber, _ := runtime.Caller(1)
+				logParams = append(logParams, "filename", fileName)
+				logParams = append(logParams, "line", lineNumber)
+		}
+		logParams = append(logParams, params...)
+		return l.logContext.Log(logParams...)
+	}
+	return nil
+}
+
+func (l *FilteredLogger) Object(obj LoggableObject) *FilteredLogger {
+
+	name := obj.GetObjectMeta().GetName()
+	uid := obj.GetObjectMeta().GetUID()
+	kind := obj.GetObjectKind()
+
+	logParams := make([]interface{}, 0)
+	logParams = append(logParams, "name", name)
+	logParams = append(logParams, "kind", kind)
+	logParams = append(logParams, "uid", uid)
+
+	l.logContext.With(logParams)
+	return l
+}
+
+func (l *FilteredLogger) With(obj ...interface{}) *FilteredLogger {
+	l.logContext.With(obj...)
+	return l
+}
+
+func (l *FilteredLogger) WithPrefix(obj ...interface{}) *FilteredLogger {
+	l.logContext.WithPrefix(obj...)
+	return l
+}
+
+func (l *FilteredLogger) SetLogLevel(filterLevel logLevel) error {
+	if (filterLevel >= DEBUG) && (filterLevel <= CRITICAL) {
+		l.filterLevel = filterLevel
+		return nil
+	}
+	return errors.New(fmt.Sprintf("Log level %d does not exist", filterLevel))
+}
+
+func (l *FilteredLogger) SetVerbosityLevel(level int) error {
+	if level >= 0 {
+		l.verbosityLevel = level
```

Comment:
Tracking these levels inside the context would mean we wanted them explicitly printed in a log line. Due to the fact that the Debug, Info, etc calls are passed in by value, they're implicitly builders, so you would not need to create a new log object just to log at a different level.

But you bring up a very good point! I will write a test case to specifically test this behavior so we ensure there's no regression in the future.